### PR TITLE
[MPS] Fix Clamp with strided outputs/inputs

### LINKED ISSUE #87348: `torch.clamp` gives wrong values for transposed inputs on MPS
### 🐛 Describe the bug

```
import numpy
import torch

torch.manual_seed(2)

shape = (2, 2)

in_ = torch.randn(shape).to('mps')
min_ = torch.randn(shape).T.to('mps')
max_ = torch.randn(shape).T.to('mps')

torched = torch.clamp(in_, min_, max_)
numpyd = numpy.clip(in_.cpu().numpy(), min_.cpu().numpy(), max_.cpu().numpy())

print("INPUTS")
print(in_)
print(min_)
print(max_)

print("OUT")
print("torch:")
print(torched.cpu().numpy())
print("numpy:")
print(numpyd)
```

Gives

```
tensor([[ 0.3923, -0.2236],
        [-0.3195, -1.2050]], device='mps:0')
tensor([[ 1.0445, -0.6332],
        [ 0.5731,  0.5409]], device='mps:0')
tensor([[-0.3919, -1.0427],
        [ 1.3186,  0.7476]], device='mps:0')
OUT
torch:
[[-0.39190584 -1.0426788 ]
 [ 1.3186173   0.747639  ]]
numpy:
[[-0.39190584 -1.0426788 ]
 [ 0.57310677  0.54094744]]
 ```
 
 This does not reproduce if the tensor isn't sent to the MPS backend, or if the matrices are not transposed
 
 Found via https://github.com/pytorch/pytorch/pull/87342

### Versions

(pytorchdev) alex@Alexs-MacBook-Pro-2 pytorch % python torch/utils/collect_env.py
Collecting environment information...
PyTorch version: 1.14.0a0+git9ceadca
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.6 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.201)
CMake version: version 3.24.2
Libc version: N/A

Python version: 3.10.4 (main, Mar 31 2022, 03:37:37) [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-12.6-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.3
[pip3] torch==1.14.0a0+git70c881b
[pip3] torchvision==0.13.1
[conda] numpy                     1.23.3                   pypi_0    pypi
[conda] numpy-base                1.23.1          py310h742c864_0
[conda] torch                     1.14.0a0+git70c881b           dev_0    <develop>
[conda] torchvision               0.13.1                   pypi_0    pyp

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### LINKED ISSUE #89673: MPS bug on `torch.transpose` and `torch.log`
### 🐛 Describe the bug

The following code uses `torch.transpose` and `torch.log` for computing the loss value.
However, it seems that the application order of the above two functions causes the different results.

```python
#!/usr/bin/env python3
import argparse
import torch
torch.random.manual_seed(31337)

parser = argparse.ArgumentParser()
parser.add_argument("--bug", help="use buggy path", action="store_true")
parser.add_argument("--device", help="specify device", type=str, default="cpu")
args = parser.parse_args()

class Diff:
  def __init__(self, model, device, use_bug):
    self.model = model
    self.device = device
    self.use_bug = use_bug
    self.lossfn = torch.nn.NLLLoss(reduction="sum")
  def forward(self, x0_indices):
    x_indices = torch.zeros((1+1, x0_indices.shape[0], self.model.length), dtype=torch.long).to(self.device)
    q = torch.zeros((1+1, x0_indices.shape[0], self.model.length, 2)).to(self.device)
    x_indices[0,] = x0_indices
    q[1,] = 0.5*torch.ones(q[1,].shape)
    x_indices[1,] = torch.distributions.Categorical(q[1,]).sample()
    return x_indices, q
  def loss(self, x0_indices):
    x_indices, q = self.forward(x0_indices)
    if self.use_bug:
      pt = torch.log(torch.transpose(self.model(x_indices[1,], 1), -2, -1))
    else:
      pt = torch.transpose(torch.log(self.model(x_indices[1,], 1)), -2, -1)
    qt = torch.log(torch.transpose(q[1,], -2, -1))
    return self.lossfn(pt, x_indices[0,])

class MLP(torch.nn.Module):
  def __init__(self, length):
    super().__init__()
    self.length = length
    self.embed_input = torch.nn.Embedding(2, 50, padding_idx=0)
    self.readouts = torch.nn.Linear(50, 2)
    self.softmax = torch.nn.Softmax(dim=-1)
  def forward(self, x_indices, t):
    x = self.embed_input(x_indices)
    x = x.reshape((x.shape[0], self.length, -1))
    return self.softmax(self.readouts(x))

x0_indices = torch.zeros((200, 20))
for i in range(x0_indices.shape[0]):
  for j in range(i%5, x0_indices.shape[1], 5):
    x0_indices[i, j] = 1

model = MLP(x0_indices.shape[1]).to(args.device)
diff = Diff(model, args.device, args.bug)

optim = torch.optim.Adam(diff.model.parameters())
for epoch in range(10000):
  loss = diff.loss(x0_indices)
  print(f"[*] epoch={epoch}: loss={loss.item():.3f}")
  if loss < 0.0:
    print(f"[-] loss is not positive")
    break
  optim.zero_grad()
  loss.backward()
  optim.step()
```

Here is my result:
```
% python3 pytorch_mps_bug.py --device cpu      
[*] epoch=0: loss=2608.710
...
[*] epoch=9999: loss=2001.556
% python3 pytorch_mps_bug.py --device cpu --bug
[*] epoch=0: loss=2608.710
...
[*] epoch=9999: loss=2001.556
% python3 pytorch_mps_bug.py --device mps      
[*] epoch=0: loss=3261.913
...
[*] epoch=9999: loss=0.016
% python3 pytorch_mps_bug.py --device mps --bug
[*] epoch=0: loss=105.605
[*] epoch=1: loss=66.850
[*] epoch=2: loss=28.175
[*] epoch=3: loss=-10.516
[-] loss is not positive
```
At least, I think loss values (`torch.nn.NLLLoss`) should not be negative value, because `torch.nn.Softmax` is applied.

In addition, the loss values after 10,000 epochs on CPU and MPS avoiding buggy path are different.
I wonder why this difference happens.

### Versions

On my Apple M2 MacBook Air:
```
% curl https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py > ./collect_env.py
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 17278  100 17278    0     0   479k      0 --:--:-- --:--:-- --:--:--  581k
% python3 ./collect_env.py 
Collecting environment information...
PyTorch version: 1.13.0a0+git6dc8fba
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.0.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.24.3
Libc version: N/A

Python version: 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:38:29) [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-13.0.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.4
[pip3] torch==1.13.0a0+git7c98e70
[conda] numpy                     1.23.4          py310h5d7c261_1    conda-forge
[conda] torch                     1.13.0a0+git7c98e70           dev_0    <develop>
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### LINKED ISSUE #94396: `torch.clamp` gives the wrong values for permuted tensors with MPS devices.
### 🐛 Describe the bug

The function `torch.clamp` gives incorrect values for tensors on MPS devices after their dimensions have been permuted.

This is likely a generalization of #87348 (https://github.com/pytorch/pytorch/issues/87348)

The following minimum example demonstrates this behavior.
```python
import torch

random_values = torch.randn(224, 224, 3, device=torch.device("mps"))
permuted_clampped_random_values = torch.clamp(values.permute(2, 0, 1), min=0, max=0.5)
print(random_values.max())
print(permuted_clampped_random_values.max())
```

### Versions

Collecting environment information...
PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.2 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.9 (main, Jan 11 2023, 09:18:18) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2

Versions of relevant libraries:
[pip3] mypy==0.981
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.5
[pip3] torch==1.13.1
[pip3] torchaudio==0.13.1
[pip3] torchvision==0.14.1
[conda] numpy                     1.23.5          py310hb93e574_0  
[conda] numpy-base                1.23.5          py310haf87e8b_0  
[conda] pytorch                   1.13.1                 py3.10_0    pytorch
[conda] torchaudio                0.13.1                py310_cpu    pytorch
[conda] torchvision               0.14.1                py310_cpu    pytorch

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by qqaatw**
Fixes #94396
Fixes #87348

1. If output is strided, we don't gather input tensors.
2. If output is not strided but min_t or max_t is strided, we make min_t or max_t contiguous.

### CODE REVIEW

**Comment by qqaatw** [comment]
@kulinseth Please take a look.

File: aten/src/ATen/native/mps/operations/TensorCompare.mm [hunk 1]
```diff
@@ -146,17 +146,30 @@ void clamp_tensor_out_mps(const Tensor& input_t,
       cachedGraph = static_cast<CachedGraph*>(tmpCachedGraph);
     }
 
-    auto inputPlaceholder = Placeholder(cachedGraph->inputTensor, input_t);
-    auto outputPlaceholder = Placeholder(cachedGraph->outputTensor, output_t);
+    bool gatherTensorData = true;
+    if (!output_t.is_contiguous() || output_t.is_view()) {
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/TensorCompare.mm:150]: This needs to be centrally fixed, rather than just for Compare operations. Here the fix is to allocate the output using empty restrided and then pass through the Strided tensor. The major concern I have is the performance implications, because we will end up doing Gather for strided inputs which in this case can be avoided if output is also strided.

Let us propose a PR which fixes it more centrally.

**Review by kulinseth — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/mps/operations/TensorCompare.mm (hunk 1 above)
Comment (qqaatw) [reply aten/src/ATen/native/mps/operations/TensorCompare.mm:150]: Actually we have been using the same fix for a while: #94260. I understood that this is not the best way to fix, but my aim was to unblock some users who have been affected by this bug for months, despite some performance implications.

Comment (qqaatw) [reply aten/src/ATen/native/mps/operations/TensorCompare.mm:150]: @kulinseth Hi, would like to know any update regarding this?

Comment (cjblocker) [reply aten/src/ATen/native/mps/operations/TensorCompare.mm:150]: If this solution is not performant enough, could we at least throw an error when we know the result will be wrong? This kind of silent incorrectness makes the mps backend nearly unusable.

Comment (kulinseth) [reply aten/src/ATen/native/mps/operations/TensorCompare.mm:150]: Hi @qqaatw, I agree, lets unblock users and we can centrally fix it later.

**Review by kulinseth — APPROVED** [review]
@qqaatw , can you please rebase the changes.
